pkg/controller/servicemesh/controlplane/addons.go

```diff
@@ -103,29 +105,34 @@ func (r *controlPlaneInstanceReconciler) patchKiali(ctx context.Context) error {
 	// credentials
 	rawPassword, err := r.getRawHtPasswd(ctx)
 	if err != nil {
-		return fmt.Errorf("could not get htpasswd required for kiali external_serivces: %s", err)
+		return reconcile.Result{}, fmt.Errorf("could not get htpasswd required for kiali external_serivces: %s", err)
 	}
 	if err := updatedKiali.Spec.SetField("external_services.grafana.auth.password", rawPassword); err != nil {
-		return fmt.Errorf("could not set external_services.grafana.auth.password in kiali CR: %s", err)
+		return reconcile.Result{}, fmt.Errorf("could not set external_services.grafana.auth.password in kiali CR: %s", err)
 	}
 	if err := updatedKiali.Spec.SetField("external_services.prometheus.auth.password", rawPassword); err != nil {
-		return fmt.Errorf("could not set external_services.prometheus.auth.password in kiali CR: %s", err)
+		return reconcile.Result{}, fmt.Errorf("could not set external_services.prometheus.auth.password in kiali CR: %s", err)
 	}
 	if err := updatedKiali.Spec.SetField("external_services.tracing.auth.password", rawPassword); err != nil {
-		return fmt.Errorf("could not set external_services.tracing.auth.password in kiali CR: %s", err)
+		return reconcile.Result{}, fmt.Errorf("could not set external_services.tracing.auth.password in kiali CR: %s", err)
 	}
 
 	// FUTURE: add support for synchronizing kiali version with control plane version
 
 	if err := r.Client.Patch(ctx, updatedKiali, client.Merge); err != nil {
 		if meta.IsNoMatchError(err) || errors.IsNotFound(err) || errors.IsGone(err) {
 			log.Info(fmt.Sprintf("skipping kiali update, %s/%s is no longer available", kiali.GetNamespace(), kiali.GetName()))
-			return nil
+			return reconcile.Result{}, nil
 		}
-		return err
+		return reconcile.Result{}, err
 	}
 
-	return nil
+	// SMCP should be reconciled if Grafana or Jaeger are enabled, but there are no routes for these services
+	if (grafanaEnabled && grafanaURL == "") || (jaegerEnabled && jaegerURL == "") {
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
```

Comment:
I'm not a fan of this approach. Have you tried watching Routes and triggering the reconcile of the relevant SMCP(s) instead? That would be the proper fix, but I understand that it could be less robust.

If we stick to the approach used here, I'd really like to see the fixed requeue time replaced with an exponential backoff, so that it takes less than 5s seconds to retry the reconcile (and takes much more time if it fails repeatedly).

Reply:
I didn't consider your approach, but it seems to be better, so I can change my PR.
To be sure I correctly understand your idea: do you suggest to add the following watches to control plane reconciler?
```
c.Watch(&source.Kind{Type: &routev1.Route{}}, enqueueRequestForSMCP, filterJaegerRoutePredicate)
c.Watch(&source.Kind{Type: &routev1.Route{}}, enqueueRequestForSMCP, filterGrafanaRoutePredicate)
```
The only downside I can see is that additional watches increase load on kube-apiserver, but on the other hand the approach with watches is consistent with the current implementation and seems to be clearer.

Reply:
I figured out that the approach with watches will not work, because Jaeger resources don't have labels `maistra.io/owner` and `maistra.io/owner-name`. Without these labels we can't determine which name and namespace to set in a reconciliation request.

Reply:
Could we add these labels to Jaeger resources? And if we can, could we add them to the previous versions as well?

Reply:
Ah yes, because we're not the ones creating those resources. How about watching the Jaeger resource? Does it contain the information we need?

Reply:
I don't know, but I will take a look at this resource once again. Maybe it's updated on route creation.

Reply:
I looked at the Jaeger resource and it doesn't contain any useful information that we could rely on. What's more, this resource is not updated when services or routes are changed, so it doesn't make sense to watch it.

Reply:
I added exponential backoff to reconciliation results.